### DISCUSSION TITLE
Introduce `@SensitiveKotlinPoetApi` annotation

### DISCUSSION THREAD

**Discussion by ZacSweers** [square/kotlinpoet#1087] [Ideas]
I've been thinking about the `TypeMirror` -> `TypeName` APIs and I think what coroutines did with `GlobalScope` is a neat middle ground: annotate them with an experimental annotation like `@SensitiveKotlinPoetApi`.

So if you opt in to using it, you're saying you Know What You're Doing™️. Then we could undeprecate them and leave them for those specific cases.

Thoughts?

**Comment by ZacSweers** [comment]
Proposal implementation https://github.com/square/kotlinpoet/pull/1088

**Comment by Egorand** [comment]
Can you please remind me if there are any legit cases when you'd want to use those APIs over the metadata ones? Only when you don't care about the lossiness of the elements API?

**Reply by ZacSweers** [reply]
Right, there are a lot of cases where they are safe to use (non-intrinsic types) and completely applicable when reading from Java sources.
